File: packages/go-sdk/operations/workflows.go
package operations

import "context"

// WorkflowsService provides workflow management operations.
type WorkflowsService struct {
	r Requester
}

// NewWorkflowsService creates a new WorkflowsService.
func NewWorkflowsService(r Requester) *WorkflowsService {
	return &WorkflowsService{r: r}
}

// List lists all workflows.
func (s *WorkflowsService) List(ctx context.Context, query map[string]string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", "/v1/workflows", query, nil, nil, &result)
	return result, err
}

// Create creates a new workflow.
func (s *WorkflowsService) Create(ctx context.Context, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/workflows", nil, nil, body, &result)
	return result, err
}

// Get gets a workflow by ID.
func (s *WorkflowsService) Get(ctx context.Context, workflowID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflows/{workflowID}", map[string]string{"workflowID": workflowID}), nil, nil, nil, &result)
	return result, err
}

// Update updates a workflow.
func (s *WorkflowsService) Update(ctx context.Context, workflowID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "PATCH", PathParams("/v1/workflows/{workflowID}", map[string]string{"workflowID": workflowID}), nil, nil, body, &result)
	return result, err
}

// Delete deletes a workflow.
func (s *WorkflowsService) Delete(ctx context.Context, workflowID string) error {
	return s.r.DoRequest(ctx, "DELETE", PathParams("/v1/workflows/{workflowID}", map[string]string{"workflowID": workflowID}), nil, nil, nil, nil)
}

// Clone clones a workflow.
func (s *WorkflowsService) Clone(ctx context.Context, workflowID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", PathParams("/v1/workflows/{workflowID}/clone", map[string]string{"workflowID": workflowID}), nil, nil, body, &result)
	return result, err
}

// DryRun validates a workflow DAG structure.
func (s *WorkflowsService) DryRun(ctx context.Context, workflowID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", PathParams("/v1/workflows/{workflowID}/dry-run", map[string]string{"workflowID": workflowID}), nil, nil, body, &result)
	return result, err
}

// Plan builds a topological workflow plan preview.
func (s *WorkflowsService) Plan(ctx context.Context, workflowID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", PathParams("/v1/workflows/{workflowID}/plan", map[string]string{"workflowID": workflowID}), nil, nil, body, &result)
	return result, err
}

// Simulate simulates workflow execution.
func (s *WorkflowsService) Simulate(ctx context.Context, workflowID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", PathParams("/v1/workflows/{workflowID}/simulate", map[string]string{"workflowID": workflowID}), nil, nil, body, &result)
	return result, err
}

// Trigger triggers a workflow run.
func (s *WorkflowsService) Trigger(ctx context.Context, workflowID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", PathParams("/v1/workflows/{workflowID}/trigger", map[string]string{"workflowID": workflowID}), nil, nil, body, &result)
	return result, err
}

// GetGraph gets the DAG visualization for a workflow run.
func (s *WorkflowsService) GetGraph(ctx context.Context, workflowRunID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflow-runs/{workflowRunID}/graph", map[string]string{"workflowRunID": workflowRunID}), nil, nil, nil, &result)
	return result, err
}

// GetGraphByWorkflowID gets the DAG visualization for a workflow.
func (s *WorkflowsService) GetGraphByWorkflowID(ctx context.Context, workflowID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflows/{workflowID}/graph", map[string]string{"workflowID": workflowID}), nil, nil, nil, &result)
	return result, err
}

// ListRuns lists runs for a workflow.
func (s *WorkflowsService) ListRuns(ctx context.Context, workflowID string, query map[string]string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflows/{workflowID}/runs", map[string]string{"workflowID": workflowID}), query, nil, nil, &result)
	return result, err
}

// ListVersions lists workflow versions.
func (s *WorkflowsService) ListVersions(ctx context.Context, workflowID string, query map[string]string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflows/{workflowID}/versions", map[string]string{"workflowID": workflowID}), query, nil, nil, &result)
	return result, err
}

// GetVersion gets a specific workflow version.
func (s *WorkflowsService) GetVersion(ctx context.Context, workflowID, versionID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflows/{workflowID}/versions/{versionID}", map[string]string{"workflowID": workflowID, "versionID": versionID}), nil, nil, nil, &result)
	return result, err
}

// GetDiff diffs two workflow versions.
func (s *WorkflowsService) GetDiff(ctx context.Context, workflowID, fromVersionID, toVersionID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflows/{workflowID}/versions/{fromVersionID}/diff/{toVersionID}", map[string]string{"workflowID": workflowID, "fromVersionID": fromVersionID, "toVersionID": toVersionID}), nil, nil, nil, &result)
	return result, err
}

// GetPolicy gets the workflow policy.
func (s *WorkflowsService) GetPolicy(ctx context.Context, projectID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflow-policies/{projectID}", map[string]string{"projectID": projectID}), nil, nil, nil, &result)
	return result, err
}

// UpsertPolicy upserts the workflow policy.
func (s *WorkflowsService) UpsertPolicy(ctx context.Context, projectID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "PUT", PathParams("/v1/workflow-policies/{projectID}", map[string]string{"projectID": projectID}), nil, nil, body, &result)
	return result, err
}

// GetExplain lists workflow step decisions.
func (s *WorkflowsService) GetExplain(ctx context.Context, workflowRunID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflow-runs/{workflowRunID}/explain", map[string]string{"workflowRunID": workflowRunID}), nil, nil, nil, &result)
	return result, err
}

// ListLabels gets workflow run labels.
func (s *WorkflowsService) ListLabels(ctx context.Context, workflowRunID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflow-runs/{workflowRunID}/labels", map[string]string{"workflowRunID": workflowRunID}), nil, nil, nil, &result)
	return result, err
}

// GetImpact gets workflow version impact.
func (s *WorkflowsService) GetImpact(ctx context.Context, workflowID, versionID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflows/{workflowID}/versions/{versionID}/impact", map[string]string{"workflowID": workflowID, "versionID": versionID}), nil, nil, nil, &result)
	return result, err
}

// ListStepsByVersion lists workflow version steps.
func (s *WorkflowsService) ListStepsByVersion(ctx context.Context, workflowID, versionID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/workflows/{workflowID}/versions/{versionID}/steps", map[string]string{"workflowID": workflowID, "versionID": versionID}), nil, nil, nil, &result)
	return result, err
}

File: packages/rust-sdk/src/operations/workflows.rs
use crate::client::StraitClient;
use crate::errors::StraitError;
use crate::http::substitute_path_params;
use std::sync::Arc;

pub struct WorkflowsService {
    client: Arc<StraitClient>,
}

impl WorkflowsService {
    pub fn new(client: Arc<StraitClient>) -> Self {
        Self { client }
    }

    pub async fn list(
        &self,
        query: Option<&[(&str, &str)]>,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("GET", "/v1/workflows", query, None, None)
            .await
    }

    pub async fn create(&self, body: serde_json::Value) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/workflows", None, None, Some(body))
            .await
    }

    pub async fn get(&self, workflow_id: &str) -> Result<serde_json::Value, StraitError> {
        let path =
            substitute_path_params("/v1/workflows/{workflowID}", &[("workflowID", workflow_id)]);
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn update(
        &self,
        workflow_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path =
            substitute_path_params("/v1/workflows/{workflowID}", &[("workflowID", workflow_id)]);
        self.client
            .do_request("PATCH", &path, None, None, Some(body))
            .await
    }

    pub async fn delete(&self, workflow_id: &str) -> Result<(), StraitError> {
        let path =
            substitute_path_params("/v1/workflows/{workflowID}", &[("workflowID", workflow_id)]);
        self.client
            .do_request_no_content("DELETE", &path, None, None, None)
            .await
    }

    pub async fn clone(
        &self,
        workflow_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/clone",
            &[("workflowID", workflow_id)],
        );
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn dry_run(
        &self,
        workflow_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/dry-run",
            &[("workflowID", workflow_id)],
        );
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn plan(
        &self,
        workflow_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/plan",
            &[("workflowID", workflow_id)],
        );
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn simulate(
        &self,
        workflow_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/simulate",
            &[("workflowID", workflow_id)],
        );
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn trigger(
        &self,
        workflow_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/trigger",
            &[("workflowID", workflow_id)],
        );
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn get_graph(&self, workflow_run_id: &str) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflow-runs/{workflowRunID}/graph",
            &[("workflowRunID", workflow_run_id)],
        );
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn get_graph_by_workflow_id(
        &self,
        workflow_id: &str,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/graph",
            &[("workflowID", workflow_id)],
        );
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn list_runs(
        &self,
        workflow_id: &str,
        query: Option<&[(&str, &str)]>,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/runs",
            &[("workflowID", workflow_id)],
        );
        self.client
            .do_request("GET", &path, query, None, None)
            .await
    }

    pub async fn list_versions(
        &self,
        workflow_id: &str,
        query: Option<&[(&str, &str)]>,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/versions",
            &[("workflowID", workflow_id)],
        );
        self.client
            .do_request("GET", &path, query, None, None)
            .await
    }

    pub async fn get_version(
        &self,
        workflow_id: &str,
        version_id: &str,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/versions/{versionID}",
            &[("workflowID", workflow_id), ("versionID", version_id)],
        );
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn get_diff(
        &self,
        workflow_id: &str,
        from_version_id: &str,
        to_version_id: &str,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/versions/{fromVersionID}/diff/{toVersionID}",
            &[
                ("workflowID", workflow_id),
                ("fromVersionID", from_version_id),
                ("toVersionID", to_version_id),
            ],
        );
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn get_policy(&self, project_id: &str) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/projects/{projectID}/workflow-policy",
            &[("projectID", project_id)],
        );
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn upsert_policy(
        &self,
        project_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/projects/{projectID}/workflow-policy",
            &[("projectID", project_id)],
        );
        self.client
            .do_request("PUT", &path, None, None, Some(body))
            .await
    }

    pub async fn get_explain(
        &self,
        workflow_run_id: &str,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflow-runs/{workflowRunID}/explain",
            &[("workflowRunID", workflow_run_id)],
        );
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn list_labels(
        &self,
        workflow_run_id: &str,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflow-runs/{workflowRunID}/labels",
            &[("workflowRunID", workflow_run_id)],
        );
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn get_impact(
        &self,
        workflow_id: &str,
        version_id: &str,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/versions/{versionID}/impact",
            &[("workflowID", workflow_id), ("versionID", version_id)],
        );
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn list_steps_by_version(
        &self,
        workflow_id: &str,
        version_id: &str,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/workflows/{workflowID}/versions/{versionID}/steps",
            &[("workflowID", workflow_id), ("versionID", version_id)],
        );
        self.client.do_request("GET", &path, None, None, None).await
    }
}
